### PR TITLE
Issue #1163 - Convert CryptHashCheck to use templates

### PR DESCRIPTION
**Pull request by Mailkov**
IA Page: https://duck.co/ia/view/crypt_hash_check

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

File: lib/DDG/Goodie/CryptHashCheck.pm [hunk 1]
```diff
@@ -24,59 +17,61 @@ code_url 'https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DD
...
-		  	my $text = sprintf qq(This is a 512 bit SHA-2/SHA-3 cryptographic hash.);
-  			my $html = sprintf qq(This is a 512 bit <a href="%s">SHA-2</a>/<a href="%s">SHA-3</a> cryptographic hash.),SHA2HREF,SHA3HREF;
+    return unless /^[0-9a-f]{32}$|^[0-9a-f]{40}$|^[0-9a-f]{56}$|^[0-9a-f]{64}$|^[0-9a-f]{96}$|^[0-9a-f]{128}$/i;
+    
+    my $chars = length $_;
+    $chars *= 4;
+ 
+    my $text = "This is a $chars bit " . $cryptohash{$chars} . " cryptographic hash."; 
+    
+    my $wiki = "https://en.wikipedia.org/wiki/";
```
Comment (moollaza) [inline lib/DDG/Goodie/CryptHashCheck.pm]: let's define this outside of the `handle`

**Comment by moollaza** [comment]
This is now live on beta! https://beta.duckduckgo.com/?q=hash+624d420035fc9471f6e16766b7132dd6bb34ea62&ia=answer

File: lib/DDG/Goodie/CryptHashCheck.pm [hunk 2]
```diff
@@ -24,59 +17,61 @@ code_url 'https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DD
...
+    
+    my @moretext;
+    my $linkname = $cryptohash{$chars}; 
+    
+    if ($cryptohash{$chars} eq "SHA-2/SHA-3") {
+        @moretext = {
```
Comment (moollaza) [inline lib/DDG/Goodie/CryptHashCheck.pm]: if you wanted to initialize this as a reference you could do:

```perl
$moretext = [{ text => "SHA-3", href => $wiki . "SHA-3" }];
```

File: lib/DDG/Goodie/CryptHashCheck.pm [hunk 3]
```diff
@@ -24,59 +17,61 @@ code_url 'https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DD
...
-  			my $html = sprintf qq(This is a 384 bit <a href="%s">SHA-2</a>/<a href="%s">SHA-3</a> cryptographic hash.),SHA2HREF,SHA3HREF;
-
-			return $text, html => $html;
-		}
 
-	my ($sha512) = /^[0-9a-f]{128}$/i;
-		if ($sha512){
-		  	my $text = sprintf qq(This is a 512 bit SHA-2/SHA-3 cryptographic hash.);
-  			my $html = sprintf qq(This is a 512 bit <a href="%s">SHA-2</a>/<a href="%s">SHA-3</a> cryptographic hash.),SHA2HREF,SHA3HREF;
+    return unless /^[0-9a-f]{32}$|^[0-9a-f]{40}$|^[0-9a-f]{56}$|^[0-9a-f]{64}$|^[0-9a-f]{96}$|^[0-9a-f]{128}$/i;
```
Comment (mintsoft) [inline lib/DDG/Goodie/CryptHashCheck.pm]: Hmm, it might be cleaner/faster to just do:
```perl
$matches = /^([0-9A-Fa-f]{32,128})$/;
my $chars = 4 * length($1);
return unless defined $cryptohash{$chars};
```

I'm not totally convinced that it's better, however all the `or`s in the regex bothers me for some reason

File: lib/DDG/Goodie/CryptHashCheck.pm (hunk 2 above)
Comment (mintsoft) [inline lib/DDG/Goodie/CryptHashCheck.pm]: I'd prefer it as a reference I think, it'd make the return cleaner

File: lib/DDG/Goodie/CryptHashCheck.pm [hunk 4]
```diff
@@ -24,59 +17,61 @@ code_url 'https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DD
...
+        };
+        $linkname = "SHA-2";
+    }
+    
+    return $text,
+    structured_answer => {
+        id => 'crypt_hash_check',
+        name => 'Answer',
+        data => {
+            title => $text,
```
Comment (mintsoft) [inline lib/DDG/Goodie/CryptHashCheck.pm]: I'd probably inline `$text` too, I can't see anywhere else it's used than here

Comment (Mailkov) [inline lib/DDG/Goodie/CryptHashCheck.pm]: @mintsoft I'm not sure what you mean, can you explain it better? Thanks!

Comment (mintsoft) [inline lib/DDG/Goodie/CryptHashCheck.pm]: @Mailkov I meant that I'd remove it as a variable, however I've just noticed that it's used in the `data` title AND the text output itself. So I'd rather not duplicate it.

TLDR; leave it as is there's no problem :)

File: lib/DDG/Goodie/CryptHashCheck.pm (hunk 3 above)
Comment (Mailkov) [inline lib/DDG/Goodie/CryptHashCheck.pm]: Your code does not work 

e.g. query `hash 1f9b59a2390bb77d2c446837d6aeab067f01b05732735f47099047cd7d3e9t85`

`Caught error: Use of uninitialized value $1 in multiplication (*) at lib/DDG/Goodie/CryptHashCheck.pm line 28.`

Comment (mintsoft) [inline lib/DDG/Goodie/CryptHashCheck.pm]: Apologies I meant:
```perl
my $matches = /^([0-9A-Fa-f]{32,128})$/ or return;
my $chars = 4 * length($1);
return unless defined $cryptohash{$chars};
```

**Comment by moollaza** [comment]
@Mailkov Thanks! This LGTM :+1:
